posts/2021/11/pydata-extensibility-vision.md

```diff
@@ -0,0 +1,363 @@
+implementation, for example, `cupyx.scipy` for GPU-accelerated SciPy functions,
+scikit-learn-intelex for Intel optimized version of scikit-learn, or cuCIM for
+GPU-accelerated scikit-image. Usually in statically typed languages function
+overloading is used to implement a function with the same name but arguments of
+different types. How can we implement something similar to functions overloading
+but in Python? Well, we could implement one central function checking the types
+of the arguments using `isinstance()`, or `issubclass()` and calling appropriate
+implementations based on these checks. This approach doesn't scale well with the
+number of inputs of different types and is not extendable by third-party
+implementers without modifying the source code. Another approach is to use
+_multiple dispatch_, where each function has multiple variants that are chosen
+based on dynamically determined types. Unfortunately, the multiple dispatch
+feature is not built-in for Python, but there's a simpler version - _single
+dispatch_. For a demo, let's begin with a tool from the Python standard library:
+`functools.singledispatch`.
+
+```python
+from functools import singledispatch
+import scipy.ndimage as _scipy_ndimage
+
+@singledispatch
+def laplace(input, output=None, mode="reflect", cval=0.0):
+    return _scipy_ndimage.laplace(input, output=output, mode=mode, cval=cval)
+
+# export `laplace` as `scipy.ndimage.laplace`
+
+# In CuPy codebase
+# from scipy.ndimage import laplace
+import cupyx.scipy.ndimage as _cupyx_ndimage
+
+@laplace.register(cupy.ndarray)
+def cupy_laplace(input, output=None, mode="reflect", cval=0.0):
+    return _cupyx_ndimage.laplace(input, output=output, mode=mode, cval=cval)
+
+# run registration code when CuPy is imported
+```
+
+We begin by defining the base `laplace` function, it is a fallback
+implementation that is going to be called for non-registered types of `input`.
+Next `laplace.register` is used to connect the implementation that is specific
+to `cupy.ndarray` with the base `laplace` function. We get a function in the
+`scipy.ndimage` module that is now compatible with CuPy arrays. Similarly, it
+can be extended to work with Dask arrays.
+
+```python
+from scipy.ndimage import laplace
+import numpy as np
+import cupy as cp
+
+numpy_img = np.random.normal(size=(256, 256))
+cupy_img = cp.array(numpy_img)
+
+print(type(laplace(numpy_img)))
+# <class 'numpy.ndarray'>
+
+print(type(laplace(cupy_img)))
+# <class 'cupy._core.core.ndarray'>
+```
+
+Since Python doesn't provide standard tools for using multiple dispatch there
+are several dedicated libraries:
+
+* _multimethod_ - pure Python implementation of multiple dispatch with caching
+  of argument types
+* _plum-dispatch_ - implementation of multiple dispatch that follows the ideas
+  from Julia
+* _uarray_ is a generic backend/multiple dispatch library developed to provide
+  NumPy-independent but similar to NumPy's dispatch functionality.
```

Comment:
This is okay but a bit vague. How about:
```suggestion
* _uarray_ is a generic backend/multiple dispatch library similar to NumPy's dispatch functionality
   except that the implementation doesn't need to be baked into the array type.
```

Reply:
Makes sense to me, adding.